### PR TITLE
[WORKFLOWS-170] Add README for documenting nf-synindex

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
Everyone loves a good README. But is this README good enough? 

You can read the nicely rendered version [here](https://github.com/Sage-Bionetworks-Workflows/nf-synindex/blob/bgrande/WORKFLOWS-170/add-readme/README.md).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,127 @@
...
+The purpose of this Nextflow workflow is to automate the process of indexing S3 objects in Synapse. These S3 objects are typically the output files from a general-purpose (_e.g._ nf-core) workflow that doesn't contain Synapse-specific steps for uploading or indexing outputs. This workflow is intended to be run after other data processing workflows.
+
+The benefits of using this workflows include:
+
+- There is a lack of data transfer and duplication because the S3 objects remain where they are.
+- The folder-like structure in S3 is reproduced in Synapse to maintain the file organization.
+
+Briefly, `nf-synindex` achieves this automation as follows:
+
+1. Update the `owner.txt` file with the authenticated user ID
```
Comment (thomasyu888) [inline README.md:14]: Is this an issue if there are different people running this pipeline.  Did we want the `owner.txt` changing?

Comment (thomasyu888) [inline README.md:14]: Is this an issue if there are different people running this pipeline.  Did we want the `owner.txt` changing?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,127 @@
...
+2. Register the Tower bucket as an external storage location
+3. Mirror the directory-like structure in S3 as folders on Synapse
+4. For each S3 object, the following steps are performed in parallel:
```
Comment (thomasyu888) [inline README.md:17]: How does this indexing deal with different versions of files?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,127 @@
...
+    s3://example-bucket/outputs/foobar_1.trimming_report.txt
+    s3://example-bucket/outputs/foobar_2.trimming_report.txt
+    s3://example-bucket/outputs/fastqc/
+    s3://example-bucket/outputs/fastqc/foobar_1_val_1_fastqc.html
+    s3://example-bucket/outputs/fastqc/foobar_1_val_1_fastqc.zip
+    s3://example-bucket/outputs/fastqc/foobar_2_val_2_fastqc.html
+    s3://example-bucket/outputs/fastqc/foobar_2_val_2_fastqc.zip
+    ```
+
+2. Prepare your Synapse configuration file to authenticate the workflow. For more details, check out the [Authentication](#authentication) section.
```
Comment (thomasyu888) [inline README.md:43]: This is pending the secrets directive correct?  Should we try to push people to use nextflow secrets over the config file?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,127 @@
...
+2. Register the Tower bucket as an external storage location
+3. Mirror the directory-like structure in S3 as folders on Synapse
+4. For each S3 object, the following steps are performed in parallel:
+   1. Compute the MD5 checksum
+   2. Register the S3 object as a file handle
+   3. Expose the file handle as a Synapse file under the corresponding Synapse folder
+5. Output a mapping between the S3 objects and the newly created Synapse files
+
+**N.B.** The workflow assumes that the S3 bucket is [configured](https://help.synapse.org/docs/Custom-Storage-Locations.2048327803.html) for Synapse.
+
```
Comment (thomasyu888) [inline README.md:24]: Same as synstage comment, would it be worth it to add a section for dependencies like tower cli?

**Review by thomasyu888 — COMMENTED** [review]
LGTM - just some comments.

File: README.md (hunk 1 above)
Comment (BrunoGrandePhD) [reply README.md:14]: The pipeline will append any missing user IDs and thus shouldn't affect the access of others to the bucket. But maybe I'm misunderstanding your concern.

File: README.md (hunk 2 above)
Comment (BrunoGrandePhD) [reply README.md:17]: I need to double-check, but I believe that the change in MD5 checksum will result in a new version being created in Synapse.

File: README.md (hunk 3 above)
Comment (BrunoGrandePhD) [reply README.md:43]: I plan on elaborating once we have deployed the secrets feature in Tower-Prod.

File: README.md (hunk 4 above)
Comment (BrunoGrandePhD) [reply README.md:24]: See my response [here](https://github.com/Sage-Bionetworks-Workflows/nf-synstage/pull/4#discussion_r836035276).

File: README.md (hunk 1 above)
Comment (thomasyu888) [reply README.md:14]: Got it, i was worried it would replace the user IDs instead of appending.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,127 @@
...
+- **`s3_prefix`**: An S3 prefix containing a set of files that need to be indexed in Synapse. Typically, this corresponds to the value given to the `outdir` parameter from an nf-core workflow run.
+
+- **`parent_id`**: The Synapse ID of a Synapse folder that will contain the indexed files and the associated folder structure.
+
+- **`synapse_config`**: (Optional) A [Synapse configuration file](https://python-docs.synapse.org/build/html/Credentials.html#use-synapseconfig) containing authentication credentials. A minimal example is included in the [Quickstart](#quickstart) section.
+
+## Known Limitations
+
+- The workflow doesn't annotate the Synapse files it creates with any metadata or provenance.
+- The workflow will index all of the objects under the given S3 prefix. There is no way to filter or skip the indexing of certain objects.
```
Comment (allaway) [inline README.md:127]:
```suggestion
- The workflow will index all of the objects under the given S3 prefix. There is no way to filter or skip the indexing of certain objects. Make sure that your `synapse_config` file is not located within a directory that will be indexed on Synapse!
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,127 @@
...
+    s3://example-bucket/outputs/foobar_2.trimming_report.txt,syn26601272
+    s3://example-bucket/outputs/fastqc/foobar_1_val_1_fastqc.html,syn26601268
+    s3://example-bucket/outputs/fastqc/foobar_1_val_1_fastqc.zip,syn26601267
+    s3://example-bucket/outputs/fastqc/foobar_2_val_2_fastqc.html,syn26601271
+    s3://example-bucket/outputs/fastqc/foobar_2_val_2_fastqc.zip,syn26601269
+    ```
+
+## Authentication
+
+Downloading files from Synapse requires the workflow to be authenticated. The workflow currently supports two authentication methods:
```
Comment (allaway) [inline README.md:107]: I'm sure I'm missing something obvious here, but isn't the purpose of this workflow to take Nextflow output files that are not on Synapse, and index them in Synapse to make them available there? In what situations would this workflow be retrieving files from Synapse?

**Review by allaway — APPROVED** [review]
LGTM! I made one minor suggested change and had one question.

File: README.md (hunk 6 above)
Comment (BrunoGrandePhD) [reply README.md:107]: Good catch! This is a copy-paste typo. I wrote the README for synstage first and didn't realize that I used a pipeline-specific term in the Authentication section.

File: README.md (hunk 5 above)
Comment (BrunoGrandePhD) [reply README.md:127]: Great suggestion! I even created an issue (#6) to build in some safeguard in the workflow itself.

**Pull request merged**